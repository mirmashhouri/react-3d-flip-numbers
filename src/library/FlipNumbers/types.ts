import type { FC } from 'react';

export type FlipDirection = 'increase' | 'decrease';

export type FlipNumberProps = {
  value: number;
  height: number;
  wrapperClassName?: string;
  numberClassName?: string;
  duration?: number;
  delay?: number;
  formatter?: (value: number) => string;
  hasShadow?: boolean;
};
export type IFlipNumbers = FC<FlipNumberProps>;
