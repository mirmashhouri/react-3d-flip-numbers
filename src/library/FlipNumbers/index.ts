export * from './FlipNumbers';
export * from './types';