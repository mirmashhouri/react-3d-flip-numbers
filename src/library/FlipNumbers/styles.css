.flip-numbers {
  display: flex;
  justify-content: center;
  align-items: center;
}
.flip-numbers .invisible {
  visibility: hidden;
}
.flip-number-container {
  display: flex;
  overflow: hidden;
  position: relative;
  z-index: 1;
  z-index: var(--z-index-main-view, 1);
  flex-direction: column;
  justify-content: center;
  align-items: center;
}
.flip-number {
  position: absolute;
  height: 100%;
  transform-style: preserve-3d;
  width: 100%;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}
.flip-number div {
  -webkit-font-smoothing: antialiased;
  backface-visibility: hidden;
}
.flip-number-fake {
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
  opacity: .85;
}
.flip-number-active {
  display: flex;
  justify-content: center;
  align-items: center;
  text-align: center;
}
.flip-shadow{
  box-shadow: black 3px 0 5px 0px;
}