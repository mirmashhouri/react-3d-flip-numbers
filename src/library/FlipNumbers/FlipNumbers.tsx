import React, { useEffect, useState } from 'react';

import { FlipNumber } from './FlipNumber';
import { IFlipNumbers } from './types';

import './styles.css';

export const FlipNumbers: IFlipNumbers = ({
  value,
  formatter = (num) => num.toString(),
  wrapperClassName = '',
  numberClassName = '',
  height,
  duration = 0.3,
  delay = 0,
  hasShadow = false,
}) => {
  const [previousValue, setPreviousValue] = useState(value);
  const [flipDirection, setFlipDirection] = useState<1 | -1>(1);

  useEffect(() => {
    setFlipDirection(value < previousValue ? 1 : -1);
    setPreviousValue(value);
  }, [value, previousValue]);

  const formattedValue = formatter(value);
  const charArray = Array.from(formattedValue);

  return (
    <section className={ `flip-numbers ${wrapperClassName}` }>
      { charArray.map((char, index) => {
        const numericValue = parseInt(char, 10);

        return !Number.isNaN(numericValue) ? (
          <FlipNumber
            key={ `flip-${index.toString()}` }
            height={ height }
            duration={ duration }
            direction={ flipDirection }
            delay={ delay }
            activeNumber={ numericValue }
            className={ numberClassName }
            hasShadow={ hasShadow }
          />
        ) : (
          <span
            className="flip-non-number"
            key={ `char-${index.toString()}` }
          >
            { char }
          </span>
        );
      }) }
    </section>
  );
};
