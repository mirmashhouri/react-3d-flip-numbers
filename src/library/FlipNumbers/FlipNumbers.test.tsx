import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';

import { FlipNumbers } from './FlipNumbers';


const component =
 <FlipNumbers
   value= { 12345 }
   formatter= { jest.fn((number) => number.toFixed(2)) }
   numberClassName= { 'number-class' }
   height= { 50 }
   duration= { 0.3 }
   delay= { 0 }
 />;

describe('FlipNumbers component', () => {
  test('renders FlipNumbers component', () => {
    const { container } = render(component);

    // wrapperClass
    expect(container.getElementsByClassName('flip-numbers')).toHaveLength(1);

    // toFixed() of fromatter add 2 number after "." 12345.00
    expect(container.getElementsByClassName('flip-number')).toHaveLength(7);
    expect(container.getElementsByClassName(component.props.numberClassName)).toHaveLength(7);

    // toFixed() of fromatter add a "."
    expect(container.getElementsByClassName('flip-non-number')).toHaveLength(1);
  });

  test('calls formatter with the correct value', () => {
    render(component);
    expect(component.props.formatter).toHaveBeenCalledWith(12345);
  });

  test('make snapshot', () => {
    const { container } = render(component);
    expect(container).toMatchSnapshot();
  });
});