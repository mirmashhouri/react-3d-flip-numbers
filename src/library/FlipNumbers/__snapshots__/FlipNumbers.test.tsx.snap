// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlipNumbers component make snapshot 1`] = `
<div>
  <section
    class="flip-numbers d-flex justify-content-center align-items-center undefined"
  >
    <div
      class="d-table text-start overflow-hidden position-relative z-index-main-view number-class"
      style="height: 50px;"
    >
      <div
        class="flip-number position-absolute h-100"
        style="transform: rotateX(-36deg); transition: transform 0.3ms 0ms ease-in-out;"
      >
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(0deg) translateZ(75px);"
        >
          0
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(36deg) translateZ(75px);"
        >
          1
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(72deg) translateZ(75px);"
        >
          2
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(108deg) translateZ(75px);"
        >
          3
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(144deg) translateZ(75px);"
        >
          4
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(180deg) translateZ(75px);"
        >
          5
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(216deg) translateZ(75px);"
        >
          6
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(252deg) translateZ(75px);"
        >
          7
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(288deg) translateZ(75px);"
        >
          8
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(324deg) translateZ(75px);"
        >
          9
        </div>
      </div>
      <div
        class="flip-number-active d-flex justify-content-center align-items-center text-center 
        invisible"
        style="height: 50px;"
      >
        1
      </div>
    </div>
    <div
      class="d-table text-start overflow-hidden position-relative z-index-main-view number-class"
      style="height: 50px;"
    >
      <div
        class="flip-number position-absolute h-100"
        style="transform: rotateX(-72deg); transition: transform 0.3ms 0ms ease-in-out;"
      >
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(0deg) translateZ(75px);"
        >
          0
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(36deg) translateZ(75px);"
        >
          1
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(72deg) translateZ(75px);"
        >
          2
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(108deg) translateZ(75px);"
        >
          3
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(144deg) translateZ(75px);"
        >
          4
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(180deg) translateZ(75px);"
        >
          5
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(216deg) translateZ(75px);"
        >
          6
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(252deg) translateZ(75px);"
        >
          7
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(288deg) translateZ(75px);"
        >
          8
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(324deg) translateZ(75px);"
        >
          9
        </div>
      </div>
      <div
        class="flip-number-active d-flex justify-content-center align-items-center text-center 
        invisible"
        style="height: 50px;"
      >
        2
      </div>
    </div>
    <div
      class="d-table text-start overflow-hidden position-relative z-index-main-view number-class"
      style="height: 50px;"
    >
      <div
        class="flip-number position-absolute h-100"
        style="transform: rotateX(-108deg); transition: transform 0.3ms 0ms ease-in-out;"
      >
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(0deg) translateZ(75px);"
        >
          0
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(36deg) translateZ(75px);"
        >
          1
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(72deg) translateZ(75px);"
        >
          2
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(108deg) translateZ(75px);"
        >
          3
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(144deg) translateZ(75px);"
        >
          4
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(180deg) translateZ(75px);"
        >
          5
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(216deg) translateZ(75px);"
        >
          6
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(252deg) translateZ(75px);"
        >
          7
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(288deg) translateZ(75px);"
        >
          8
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(324deg) translateZ(75px);"
        >
          9
        </div>
      </div>
      <div
        class="flip-number-active d-flex justify-content-center align-items-center text-center 
        invisible"
        style="height: 50px;"
      >
        3
      </div>
    </div>
    <div
      class="d-table text-start overflow-hidden position-relative z-index-main-view number-class"
      style="height: 50px;"
    >
      <div
        class="flip-number position-absolute h-100"
        style="transform: rotateX(-144deg); transition: transform 0.3ms 0ms ease-in-out;"
      >
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(0deg) translateZ(75px);"
        >
          0
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(36deg) translateZ(75px);"
        >
          1
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(72deg) translateZ(75px);"
        >
          2
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(108deg) translateZ(75px);"
        >
          3
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(144deg) translateZ(75px);"
        >
          4
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(180deg) translateZ(75px);"
        >
          5
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(216deg) translateZ(75px);"
        >
          6
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(252deg) translateZ(75px);"
        >
          7
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(288deg) translateZ(75px);"
        >
          8
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(324deg) translateZ(75px);"
        >
          9
        </div>
      </div>
      <div
        class="flip-number-active d-flex justify-content-center align-items-center text-center 
        invisible"
        style="height: 50px;"
      >
        4
      </div>
    </div>
    <div
      class="d-table text-start overflow-hidden position-relative z-index-main-view number-class"
      style="height: 50px;"
    >
      <div
        class="flip-number position-absolute h-100"
        style="transform: rotateX(-180deg); transition: transform 0.3ms 0ms ease-in-out;"
      >
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(0deg) translateZ(75px);"
        >
          0
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(36deg) translateZ(75px);"
        >
          1
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(72deg) translateZ(75px);"
        >
          2
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(108deg) translateZ(75px);"
        >
          3
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(144deg) translateZ(75px);"
        >
          4
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(180deg) translateZ(75px);"
        >
          5
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(216deg) translateZ(75px);"
        >
          6
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(252deg) translateZ(75px);"
        >
          7
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(288deg) translateZ(75px);"
        >
          8
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(324deg) translateZ(75px);"
        >
          9
        </div>
      </div>
      <div
        class="flip-number-active d-flex justify-content-center align-items-center text-center 
        invisible"
        style="height: 50px;"
      >
        5
      </div>
    </div>
    <span
      class="flip-non-number"
    >
      .
    </span>
    <div
      class="d-table text-start overflow-hidden position-relative z-index-main-view number-class"
      style="height: 50px;"
    >
      <div
        class="flip-number position-absolute h-100"
        style="transform: rotateX(0deg); transition: transform 0.3ms 0ms ease-in-out;"
      >
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(0deg) translateZ(75px);"
        >
          0
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(36deg) translateZ(75px);"
        >
          1
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(72deg) translateZ(75px);"
        >
          2
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(108deg) translateZ(75px);"
        >
          3
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(144deg) translateZ(75px);"
        >
          4
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(180deg) translateZ(75px);"
        >
          5
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(216deg) translateZ(75px);"
        >
          6
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(252deg) translateZ(75px);"
        >
          7
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(288deg) translateZ(75px);"
        >
          8
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(324deg) translateZ(75px);"
        >
          9
        </div>
      </div>
      <div
        class="flip-number-active d-flex justify-content-center align-items-center text-center 
        invisible"
        style="height: 50px;"
      >
        0
      </div>
    </div>
    <div
      class="d-table text-start overflow-hidden position-relative z-index-main-view number-class"
      style="height: 50px;"
    >
      <div
        class="flip-number position-absolute h-100"
        style="transform: rotateX(0deg); transition: transform 0.3ms 0ms ease-in-out;"
      >
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(0deg) translateZ(75px);"
        >
          0
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(36deg) translateZ(75px);"
        >
          1
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(72deg) translateZ(75px);"
        >
          2
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(108deg) translateZ(75px);"
        >
          3
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(144deg) translateZ(75px);"
        >
          4
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(180deg) translateZ(75px);"
        >
          5
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(216deg) translateZ(75px);"
        >
          6
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(252deg) translateZ(75px);"
        >
          7
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(288deg) translateZ(75px);"
        >
          8
        </div>
        <div
          class="flip-number-fake position-absolute d-flex justify-content-center align-items-center text-center 
            "
          style="height: 50px; transform: rotateX(324deg) translateZ(75px);"
        >
          9
        </div>
      </div>
      <div
        class="flip-number-active d-flex justify-content-center align-items-center text-center 
        invisible"
        style="height: 50px;"
      >
        0
      </div>
    </div>
  </section>
</div>
`;
